test/chart/chart_test.go

```diff
@@ -0,0 +1,282 @@
+		os.Exit(1)
+	}
+
+	result = m.Run()
+
+	cancel()
+	wg.Wait()
+	os.Exit(result)
+}
+
+func TestChart_upgradeCRDs(t *testing.T) {
+	operatorImageRepo := os.Getenv("IMAGE_TAG_BASE")
+	if operatorImageRepo == "" {
+		require.Fail(t, "IMAGE_TAG_BASE is not set")
+	}
+	operatorImageTag := os.Getenv("VERSION")
+	if operatorImageTag == "" {
+		require.Fail(t, "VERSION is not set")
+	}
+
+	startChartVersion := os.Getenv("TEST_START_CHART_VERSION")
+	if startChartVersion == "" {
+		startChartVersion = "0.2.0"
+	}
+
+	image := fmt.Sprintf("%s:%s", operatorImageRepo, operatorImageTag)
+	releaseName := strings.Replace(strings.ToLower(t.Name()), "_", "-", -1)
+	ctx := context.Background()
+	t.Cleanup(func() {
+		assert.NoError(t, uninstallVSO(t, ctx,
+			"--wait",
+			"--namespace", vsoNamespace,
+			releaseName,
+		))
+	})
+
+	require.NoError(t, runKind(t, ctx,
+		"load", "docker-image", image,
+		"--name", kindClusterName,
+	))
+
+	require.NoError(t, installVSO(t, ctx,
+		"--wait",
+		"--create-namespace",
+		"--namespace", vsoNamespace,
+		// picking a lower version to ensure we see some CRD changes.
+		"--version", startChartVersion,
+		releaseName,
+		"hashicorp/vault-secrets-operator",
+	))
+
+	var origCRDs apiextensionsv1.CustomResourceDefinitionList
+	require.NoError(t, client.List(ctx, &origCRDs))
+
+	origCRDsByName := map[string]apiextensionsv1.CustomResourceDefinition{}
+	for _, o := range origCRDs.Items {
+		origCRDsByName[o.Name] = o
+	}
+
+	require.NoError(t, upgradeVSO(t, ctx,
+		"--wait",
+		"--namespace", vsoNamespace,
+		"--set", fmt.Sprintf("controller.manager.image.repository=%s", operatorImageRepo),
+		"--set", fmt.Sprintf("controller.manager.image.tag=%s", operatorImageTag),
+		releaseName,
+		chartPath,
+	))
+
+	crdsDir := filepath.Join(chartPath, "crds")
+	wantCRDs, err := utils.LoadCRDsFromDir(crdsDir)
+	require.NoError(t, err, "failed to load CRDs from %q", crdsDir)
+	require.Greater(t, len(wantCRDs), 0, "no CRDs found in %q", crdsDir)
+
+	var updatedCRDs apiextensionsv1.CustomResourceDefinitionList
+	require.NoError(t, client.List(ctx, &updatedCRDs))
+	assert.NotEqual(t, origCRDs.Items, updatedCRDs.Items)
+
+	for _, wantCRD := range wantCRDs {
+		var updatedCRD apiextensionsv1.CustomResourceDefinition
+		require.NoError(t, client.Get(ctx, ctrlclient.ObjectKeyFromObject(&wantCRD), &updatedCRD))
+		if wantCRD.Spec.Conversion == nil {
+			updatedCRD.Spec.Conversion = nil
+		}
+
+		if o, ok := origCRDsByName[wantCRD.Name]; ok {
+			assert.Greater(t, updatedCRD.Generation, o.Generation)
+			assert.Equal(t, o.UID, updatedCRD.UID)
+		} else {
+			assert.Equal(t, int64(1), updatedCRD.Generation)
+		}
+
+		assert.Equal(t, wantCRD.Spec, updatedCRD.Spec, "CRD %q .spec mismatch", wantCRD.Name)
+		assert.Equal(t, wantCRD.Labels, updatedCRD.Labels, "CRD %q .metadata.labels mismatch", wantCRD.Name)
+		assert.Equal(t, wantCRD.Annotations, updatedCRD.Annotations, "CRD %q .metadata.annotations mismatch", wantCRD.Name)
+	}
```

Comment:
Might be good to compare `wantCRD.Status` as well if possible.

Reply:
Fixed in 9b5c3a12a05ea0cce3a8938cbc98bbc4c2662de4